src/imports.rs

```diff
@@ -1,66 +1,61 @@
+#![cfg(target_arch = "wasm32")]
+
+use std::ffi::c_void;
 use std::vec::Vec;
 
-pub trait Storage {
-    fn get(&self, key: &[u8]) -> Option<Vec<u8>>;
-    fn set(&mut self, key: &[u8], value: &[u8]);
-}
+use crate::memory::{alloc, build_slice, consume_slice, Slice};
+use crate::storage::Storage;
 
-#[cfg(target_arch = "wasm32")]
-pub use wasm::ExternalStorage;
+// this is the buffer we pre-allocate in get - we should configure this somehow later
+static MAX_READ: usize = 2000;
 
-#[cfg(target_arch = "wasm32")]
-mod wasm {
-    use super::*;
-    use crate::memory::{alloc, build_slice, consume_slice, Slice};
-    use std::ffi::c_void;
+extern "C" {
+    // both take an opaque database ref that can be used by the environment to determine which
+    // substore to allow read/writes from
+    fn c_read(key: *const c_void, value: *mut c_void) -> i32;
+    fn c_write(key: *const c_void, value: *mut c_void);
+}
 
-    // this is the buffer we pre-allocate in get - we should configure this somehow later
-    static MAX_READ: usize = 2000;
+pub struct ExternalStorage {}
 
-    extern "C" {
-        // both take an opaque database ref that can be used by the environment to determine which
-        // substore to allow read/writes from
-        fn c_read(key: *const c_void, value: *mut c_void) -> i32;
-        fn c_write(key: *const c_void, value: *mut c_void);
+impl ExternalStorage {
+    pub fn new() -> ExternalStorage {
+        ExternalStorage {}
     }
+}
 
-    pub struct ExternalStorage {}
-
-    impl ExternalStorage {
-        pub fn new() -> ExternalStorage {
-            ExternalStorage {}
+impl Storage for ExternalStorage {
+    fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+        let key = build_slice(key);
+        let key_ptr = &*key as *const Slice as *const c_void;
+        let value = alloc(MAX_READ);
+
+        let read = unsafe { c_read(key_ptr, value) };
+        if read < 0 {
+            // TODO: try to read again with larger amount
+            panic!("needed to read more data")
+        } else if read == 0 {
+            return None;
         }
-    }
 
-    impl Storage for ExternalStorage {
-        fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
-            unsafe {
-                let key = build_slice(key);
-                let key_ptr = &*key as *const Slice as *const c_void;
-                let value = alloc(MAX_READ);
-                let read = c_read(key_ptr, value);
-                let mut data = consume_slice(value);
-                if read < 0 {
-                    // TODO: try to read again with larger amount
-                    panic!("needed to read more data")
-                } else if read == 0 {
-                    return None;
-                } else {
-                    data.truncate(read as usize);
-                    return Some(data);
-                }
+        let mut data = unsafe {
+            match consume_slice(value) {
```

Comment:
⛳️ 
```
let mut data = unsafe { consume_slice().ok() };
```

Reply:
⛳️ 
`unsafe { consume_slice().ok() }.map(|data| data.truncate(read as usize)`

Reply:
Very nice.
Both the chain, and using `.ok()` to do the Result to Option conversion.

Hole in one